cpp/src/reshape/interleave_columns.cu

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/types.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include "cudf/utilities/bit.hpp"
+
+namespace cudf {
+
+namespace experimental {
+
+namespace detail {
+
+template<typename TElement>
+struct interleave_columns_selector
+{
+    table_device_view in;
+
+    TElement __device__ operator()(size_type i)
+    {
+        auto col_num = i % in.num_columns();
+        auto row_num = i / in.num_columns();
+        column_device_view in_col = in.column(col_num);
+        return in_col.element<TElement>(row_num);
+    }
+};
+
+struct interleave_columns_validity_selector
+{
+    table_device_view in;
+    size_type out_row_count;
+
+    bitmask_type __device__ operator()(size_type i)
+    {
+        bitmask_type out = 0b00000000000000000000000000000000;
+
+        const auto num_bits = cudf::detail::size_in_bits<bitmask_type>();
+
+        for (size_type bit = 0; bit < num_bits; bit++) {
```

Comment:
I think this is going to be considerably slower than a `ballot_sync` kernel.

[`valid_if`](https://github.com/rapidsai/cudf/blob/branch-0.12/cpp/include/cudf/detail/valid_if.cuh#L91) could be used to alleviate the need for writing a new kernel here.

Reply:
Been focusing on quantiles, so I haven't worked on this since we talked about it. I intend to use a custom kernel and potentially make some kernel helper methods that can be shared across bitmask kernels.